### PR TITLE
feat: Add EnableFunctionDefaultPermissions property to HTTP Api

### PR DESCRIPTION
**Pull request by aaythapa**
### Issue #, if available

#2933 

### Description of changes

Added the `EnableFunctionDefaultPermissions` (related pr: #2885) to SAM Http Api. By default, HTTP Api doesn't create the permissions resource that allows the API to invoke the lambda authorizer so this property will be False by default. When its set to True it'll create the permissions resource.

### Description of how you validated changes

Ran a test with the property set to false and without the property. Compared outputs and made sure only change was that the permissions resource was added. Deployed a template with SAM Http Api with the permission resource and made sure the API could invoke the authorizer. 

### Checklist

- [x] Adheres to the [development guidelines](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#development-guidelines)
- [x] Add/update [transform tests](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#unit-testing-with-multiple-python-versions)
    - [x] Using correct values
    - [x] Using wrong values
- [ ] Add/update [integration tests](https://github.com/aws/serverless-application-model/blob/develop/INTEGRATION_TESTS.md)

### Examples?

Please reach out in the comments if you want to add an example. Examples will be 
added to `sam init` through [aws/aws-sam-cli-app-templates](https://github.com/aws/aws-sam-cli-app-templates).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: samtranslator/model/api/http_api_generator.py [hunk 1]
```diff
@@ -511,6 +513,61 @@ def _add_tags(self) -> None:
...
+        """Constructs and returns the Lambda Permission resource allowing the Authorizer to invoke the function.
+
+        :returns: the permission resource
+        :rtype: model.lambda_.LambdaPermission
+        """
+
+        partition = ArnGenerator.get_partition_name()
+        resource = "${__ApiId__}/authorizers/*"
+        source_arn = fnSub(
+            ArnGenerator.generate_arn(partition=partition, service="execute-api", resource=resource),
```
Comment (hoffa) [inline samtranslator/model/api/http_api_generator.py:528]: We shouldn’t hardcode the partition or linters will complain. Better to use the pseudo-parameter.

**Review by ssenchenko — APPROVED** [review]

File: samtranslator/model/sam_resources.py [hunk 2]
```diff
@@ -1388,15 +1388,13 @@ def to_cloudformation(self, **kwargs):  # type: ignore[no-untyped-def]
...
         )
 
-        (
-            http_api,
-            stage,
-            domain,
-            basepath_mapping,
-            route53,
-        ) = api_generator.to_cloudformation(kwargs.get("route53_record_set_groups", {}))
+        (http_api, stage, domain, basepath_mapping, route53, permissions) = api_generator.to_cloudformation(
```
Comment (hoffa) [inline samtranslator/model/sam_resources.py:1391]: nit: better to keep the previous formatting (add trailing comma), it's a misleadingly large diff for a tiny change

File: samtranslator/model/apigatewayv2.py [hunk 3]
```diff
@@ -166,6 +175,11 @@ def _validate_input_parameters(self) -> None:
                 self.api_logical_id, "EnableSimpleResponses must be defined only for Lambda Authorizer."
             )
 
+        if self.disable_function_default_permissions is not None and not authorizer_type == "REQUEST":
```
Comment (hoffa) [inline samtranslator/model/apigatewayv2.py:178]:
```suggestion
        if self.disable_function_default_permissions is not None and authorizer_type != "REQUEST":
```

File: samtranslator/model/api/http_api_generator.py [hunk 4]
```diff
@@ -511,6 +513,60 @@ def _add_tags(self) -> None:
...
+    ) -> LambdaPermission:
+        """Constructs and returns the Lambda Permission resource allowing the Authorizer to invoke the function.
+
+        :returns: the permission resource
+        :rtype: model.lambda_.LambdaPermission
+        """
+
+        resource = "${__ApiId__}/authorizers/*"
+        source_arn = fnSub(
+            ArnGenerator.generate_arn(partition="${AWS::Region}", service="execute-api", resource=resource),
```
Comment (hoffa) [inline samtranslator/model/api/http_api_generator.py:527]: > partition="${AWS::Region}"

🤨

Comment (aaythapa) [reply samtranslator/model/api/http_api_generator.py]: brain fart

File: samtranslator/model/apigatewayv2.py [hunk 5]
```diff
@@ -115,6 +117,13 @@ def __init__(  # type: ignore[no-untyped-def] # noqa: too-many-arguments
         if authorizer_type == "REQUEST":
             self._validate_lambda_authorizer()
 
+        if disable_function_default_permissions is not None:
+            sam_expect(
+                disable_function_default_permissions,
+                api_logical_id,
+                f"Authorizers.{name}.DisableFunctionDefaultPermissions",
+            ).to_be_a_bool()
```
Comment (hoffa) [inline samtranslator/model/apigatewayv2.py:125]: Would be clearer to add a `to_be_a_bool_or_none`.

**Review by ssenchenko — APPROVED** [review]

File: samtranslator/model/apigatewayv2.py (hunk 5 above)
Comment (aaythapa) [reply samtranslator/model/apigatewayv2.py:125]: need to refactor sam_expect a bit more so going to make this a separate PR

File: tests/model/test_api_v2.py [hunk 6]
```diff
@@ -168,6 +168,21 @@ def test_create_authorizer_fails_with_enable_simple_responses_non_lambda(self):
...
+                api_logical_id="logicalId",
+                name="authName",
+                jwt_configuration={"config": "value"},
+                authorization_scopes=["scope1", "scope2"],
+                enable_function_default_permissions=True,
+            )
+        self.assertEqual(
+            e.value.message,
+            "Resource with id [logicalId] is invalid. "
+            + "EnableFunctionDefaultPermissions must be defined only for Lambda Authorizer.",
```
[the remark this answers is not in the document]
Comment (aaythapa) [reply tests/model/test_api_v2.py:183]: added error test

File: samtranslator/model/api/http_api_generator.py [hunk 7]
```diff
@@ -511,6 +513,60 @@ def _add_tags(self) -> None:
...
+    def _get_permission(
+        self, authorizer_name: str, authorizer_lambda_function_arn: str, api_arn: str
+    ) -> LambdaPermission:
+        """Constructs and returns the Lambda Permission resource allowing the Authorizer to invoke the function.
+
+        :returns: the permission resource
+        :rtype: model.lambda_.LambdaPermission
+        """
+
+        resource = "${__ApiId__}/authorizers/*"
```
[the remark this answers is not in the document]
Comment (hoffa) [reply samtranslator/model/api/http_api_generator.py:525]: Specifically, is it different from `authorizer_name`?

[the remark this answers is not in the document]
Comment (aaythapa) [reply samtranslator/model/api/http_api_generator.py:525]: Gives any api ability to invoke any authorizer. It was the same permissions as SAM Api but should it be stricter?

[the remark this answers is not in the document]
Comment (aaythapa) [reply samtranslator/model/api/http_api_generator.py:525]: Just checked adding `authorizer_name` to the resource won't give the right permissions to invoke the authorizer.

**Pull request merged**